### ISSUE TITLE
threads - in Python and ramification on Terrier data structures.

### ISSUE THREAD

**Issue by cmacdonald** [terrier-org/pyterrier#3]

**Comment by seanmacavaney** [comment]
It would be nice to figure this out so BatchRetrieve could process the queries in parallel.

**Comment by cmacdonald** [comment]
This is about Terrier's internals:
 - Lexicon is not threadsafe
 - BitFileBuffered is not threadsafe
 - Matching is not threadsafe

There is some work in terrier-concurrent that addresses some of this, by appropriate synchronisation of structures. However, it currently does a lot of locking that isnt necessary (essentially a lock around Matching); a workaround would be a new Matching instance for each query, but that would have overheads for some settings (e.g. loading expensive features).
 
The long-term aim is to merge @tonellotto 's terrier-micro (with EF data structures) back into mainstream Terrier. All that can be the focus for now is fixes to to terrier-concurrent.

terrier-concurrent can be used by simply prefixing the indexref location with "concurrent:". YMMV.

**Comment by seanmacavaney** [comment]
Awesome- thanks for the info! I'll give it a try.

**Comment by cmacdonald** [comment]
to be clear, unfortunately, IIRC the lock around Matching probably means you wont get much of a speedup..

**Comment by cmacdonald** [comment]
I committed to https://github.com/terrier-org/pyterrier/tree/tr_concurrent a pass at what this might look like, but terrier-concurrent isn't sufficient:

```
['java.lang.RuntimeException: Problem looking for doclength for document 1 -- docid out of bounds, possible (concurrent?) decompression error', 
'org.terrier.structures.postings.bit.BasicIterablePosting.getDocumentLength(BasicIterablePosting.java:87)', 
'org.terrier.matching.models.WeightingModel.score(WeightingModel.java:124)', 
'org.terrier.matching.PostingListManager.score(PostingListManager.java:460)', 
'org.terrier.matching.daat.Full.assignScore(Full.java:209)', 'org.terrier.matching.daat.Full.match(Full.java:133)', 
'org.terrier.querying.LocalManager$ApplyLocalMatching.process(LocalManager.java:522)', 
'org.terrier.querying.LocalManager.runSearchRequest(LocalManager.java:916)
```

The hack workaround is to open an index for each thread (or perhaps create a cache of them). I would avoid this if possible, as Terrier indices consume significant memory.

**Comment by seanmacavaney** [comment]
Gotcha, thanks.

**Comment by cmacdonald** [comment]
For those interested in speeding up retrieval of PyTerrier, see https://pyterrier.readthedocs.io/en/latest/parallel.html

**Comment by cmacdonald** [comment]
Work is ongoing in the [rentrant branch](https://github.com/terrier-org/terrier-core/tree/reentrant) of terrier-core and [tr_retrant branch](https://github.com/terrier-org/pyterrier/tree/tr_rentrant) of pyterrier. Mean response times for the MSMARCO passage ranking corpus with different numbers of threads:

| threads | MRT  |
|---------|------|
| 1             | 146  |
| 2            | 99   |
| 3            |  118 |
| 4            | 155  |
| 5            | 189  |

Observations: 
 - this is for on-disk indices
 - we need to make it easier in PyTerrier to load indices entirely into memory
 - we are not seeing massive benefits for increased threads, presumably due to disk IO contention
 - using the inflate compression for CompressingMetaIndex is a costly overhead
 - if the metaindex only contains docnos, compression may by superfluous
 - replacing inflate compression with zstd compression would bring decompression speed benefits - see thread https://twitter.com/craig_macdonald/status/1393717764951527425

**Comment by seanmacavaney** [comment]
I took a moderately deep dive into document metadata compression a while back and generally found [lz4](https://github.com/lz4/lz4) compression to offer the best decompression speeds, with a reasonable compression speed and a good compression ratio. It's what I ended up using in irds.

I didn't try zstd specifically, though. Looks like by [zstd's benchmarks](https://github.com/facebook/zstd#benchmarks), lz4 can't match zstd's best compression ratio, but it's still much faster in terms of decompression speed (~twice as fast at the same compression ratio).
